app/src/main/java/com/duckduckgo/app/history/HistoryRepository.kt

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2024 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.history
+
+import com.duckduckgo.app.history.store.HistoryDao
+import com.duckduckgo.common.utils.CurrentTimeProvider
+import io.reactivex.Single
+
+interface HistoryRepository {
+    fun getHistoryObservable(): Single<List<HistoryEntry>>
+    fun saveToHistory(
+        url: String,
+        title: String?,
+        query: String?,
+        isSerp: Boolean,
+    )
+}
+
+class RealHistoryRepository(
+    private val historyDao: HistoryDao,
+    private val currentTimeProvider: CurrentTimeProvider,
+) : HistoryRepository {
+
+    private var cachedHistoryEntries: List<HistoryEntry>? = null
+
+    override fun getHistoryObservable(): Single<List<HistoryEntry>> {
+        return if (cachedHistoryEntries != null) {
+            Single.just(cachedHistoryEntries)
+        } else {
+            fetchAndCacheHistoryEntries()
+        }
+    }
+
+    override fun saveToHistory(
+        url: String,
+        title: String?,
+        query: String?,
+        isSerp: Boolean,
+    ) {
+        historyDao.updateOrInsertVisit(url, title ?: "", query, isSerp, currentTimeProvider.currentTimeMillis())
+        fetchAndCacheHistoryEntries()
```

Comment:
I assume this will change so we don't query all entries on every insert, as it's just part of the MVP. just flagging in case we missed this part.

Reply:
I want to make sure what I have in memory is the same I have in the database, but I think using an observable is a better way to go here. Alternatively, I could just make updateOrInsertVisit return the inserted item and update the cached list myself.

Reply:
Yeah, both approaches should work here.

Reply:
I tested the observable approach better, but unfortunately, every time a new query is submitted, the observable is disposed, so the approach is useless, as we are effectively not caching anything. As for inserting the item and updating the cache myself, even though it might be a more efficient approach, it could lead to inconsistencies easily, as we'd need to handle duplicates twice (for database and cache)

Reply:
Closing this thread as we've already discussed the approach